geopandas/sindex.py

```diff
@@ -70,6 +91,150 @@ def __init__(self, geometry):
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+
+            # handle invalid predicates
+            if predicate not in self.valid_query_predicates:
+                raise ValueError(
+                    "Got `predicate` = `{}`, `predicate` must be one of {}".format(
+                        predicate, self.valid_query_predicates
+                    )
+                )
+
+            # handle empty / invalid geometries
+            if geometry is None:
+                # return an empty integer array, similar to pygeys.STRtree.query.
+                return np.array([], dtype=np.intp)
+
+            if not isinstance(geometry, BaseGeometry):
+                raise TypeError(
+                    "Got `geometry` of type `{}`, `geometry` must be ".format(
+                        type(geometry)
+                    )
+                    + "a shapely geometry."
+                )
+
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # query tree
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_idx = list(self.intersection(bounds, objects=False))
+
+            if not tree_idx:
+                return np.array([], dtype=np.intp)
+
+            # Check predicate
+            # When possible, we want to use cached prepared geometries
+            # from the tree to compare tree_geom.predicate(input_geom)
+            # If this is not possible, try prep the input geometry
+            # and compare that against results from the tree.
+            # Since only certain predicates support prepared geometries,
+            # the fallback is to check non-prepared geometries.
+            if predicate in ("intersects", "within"):
+                # For these two predicates, we compare tree_geom.predicate(input_geom)
+                res = []
+                if predicate == "within":
+                    # The user asked for input_geom.within(tree_geom)
+                    # but since within and contains are inverse,
+                    # we can do the same test as tree_geometry.contains(input_geometry)
+                    # to use cached prepared tree geometry
+                    predicate = "contains"
+                for i in tree_idx:
+                    if self._prepared_geometries[i] is None:
+                        # if not already prepared, prepare and cache
+                        self._prepared_geometries[i] = prep(self._geometries[i])
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
+                        res.append(i)
+                tree_idx = res
+            elif predicate is not None:
+                # For the remaining predicates,
+                # we compare input_geom.predicate(tree_geom)
+                if len(tree_idx) > 1 and predicate == "contains":
```

Comment:
```suggestion
                if predicate == "contains":
```

I don't think this check is needed. Even for a single predicate call, preparing the geom can already be beneficial

Reply:
Okay, thank you for confirming that. I wasn't sure if the threshold for which preping was faster was 1 comparison, 5 comparisons or 10 comparisons. If prepping is faster for even 1 comparison I agree, the always prep is better.